support/types.go

```diff
@@ -0,0 +1,65 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package support
+
+type ClusterOpts func(c E2EClusterProvider)
+
+type E2EClusterProvider interface {
+	// WithName is used to configure the cluster Name that should be used while setting up the cluster. Might
+	// Not apply for all providers.
+	WithName(name string) E2EClusterProvider
+
+	// WithVersion helps you override the default version used while using the cluster provider.
+	// This can be useful in providing a mechanism to the end users where they want to test their
+	// code against a certain specific version of k8s that is not the default one configured
+	// for the provider
+	WithVersion(version string) E2EClusterProvider
+
+	// WithOpts provides a way to customize the options that can be used while setting up the
+	// cluster using the providers such as kind or kwok or anything else. These helpers can be
+	// leveraged to setup arguments or configuration values that can be provided while performing
+	// the cluster bring up
+	WithOpts(opts ...ClusterOpts) E2EClusterProvider
+
+	// Create Provides an interface to start the cluster creation workflow using the selected provider
+	Create(args ...string) (string, error)
+
+	CreateWithConfig(kindConfigFile string) (string, error)
+
+	// GetKubeconfig provides a way to extract the kubeconfig file associated with the cluster in question
+	// using the cluster provider native way
+	GetKubeconfig() string
+
+	// GetKubeCtlContext is used to extract the kubectl context to be used while performing the operation
+	GetKubeCtlContext() string
+
+	// ExportLogs is used to export the cluster logs via the cluster provider native workflow. This
+	// can be used to export logs from the cluster after test failures for example to analyze the test
+	// failures better after the fact.
+	ExportLogs(dest string) error
+
+	// Destroy is used to cleanup a cluster brought up as part of the test workflow
+	Destroy() error
+
+	// LoadDockerImage is used to load a set of Docker images to the cluster via the cluster provider native workflow
+	LoadDockerImage(image string) error
+
+	// LoadImageArchive is used to provide a mechanism where a tar.gz archive containing the docker images used
+	// by the services running on the cluster can be imported and loaded into the cluster prior to the execution of
+	// test if required.
+	LoadImageArchive(archivePath string) error
```

Comment:
I'm still trying to get familiar with `kwok` but I'm not sure if there is a concept of this so these may be specific enough to `kind` to leave off? Imagining a (albeit complex) `kubeadm` or OpenShift (I guess there is `oc import-image`) provider, I'm not sure if these would apply either. Curious to hear your thoughts!

Reply:
+1, did not find an option in `kwokctl` to do this.

Reply:
@vladimirvivien  @jbpratt @reetasingh  

I totally agree that not all providers need to have this capabilities or will provide these capabilities. I kept that under this just to simplify things and avoid having to type assert things and then invoke handlers selectively in some places if we decide to use these functions in the future. The providers that do not need to support this can just have a dummy handler defined that is a no-op. Or for cases of `kubeadm` for example, it can `docker/crictl pull` the image to the cluster as part of the load if someone decides to take that route. 

Is there a real advantages of taking this out of this type ?